### PR TITLE
Fix reloading page issue when user make Stake S3F/ Unstake S3F actions on page “Earn SNOB”  (deprecated)

### PR DESCRIPTION
**Pull request by thongpt-codelink**
## Status
**Done**

## Description
- The logic for fetching information and rendering the poolS3F is now separated, no need to wait until other fetching resolved.
- Instead of reloading the page when the action is submitted, fetching only relevant information and re-rendering the element that the action interact.

https://user-images.githubusercontent.com/83624863/117252489-38f3bd00-ae70-11eb-9d71-bd6709562fa4.mov

### CODE REVIEW

File: src/static/js/earn.js [hunk 1]
```diff
@@ -73,6 +74,88 @@ async function main() {
   const LINK_AVAX_TVL = "https://info.pangolin.exchange/#/account/0x974Ef0bDA58C81F3094e124f530eF34fe70dc103"
   const USDT_AVAX_TVL = "https://info.pangolin.exchange/#/account/0x74dB28797957a52a28963F424dAF2B10226ba04C"
 
+  const stakingContract_stake = async ({
+    STAKING_ABI, STAKING_ADDR, S3F_ADDRESS, App, STAKING_CONTRACT, SNOB_TOKEN, S3F_TOKEN, renderPoolS3F
+  }) => {
+    const signer = App.provider.getSigner()
+    console.log(signer)
```
Comment (huynvk) [inline src/static/js/earn.js]: Please remove the unneccessary logs

File: src/static/js/earn.js [hunk 2]
```diff
@@ -73,6 +74,88 @@ async function main() {
...
+    const STAKING_TOKEN = new ethers.Contract(S3F_ADDRESS, ERC20_ABI, signer)
+    console.log(STAKING_TOKEN)
+    const CHEF_CONTRACT = new ethers.Contract(STAKING_ADDR, STAKING_ABI, signer)
+    console.log(CHEF_CONTRACT)
+    const currentTokens = await STAKING_TOKEN.balanceOf(App.YOUR_ADDRESS)
+    console.log(currentTokens)
+    const allowedTokens = await STAKING_TOKEN.allowance(App.YOUR_ADDRESS, STAKING_ADDR)
+    console.log(allowedTokens)
+    let allow = Promise.resolve()
+    if (allowedTokens / 1e18 == 0) {
```
Comment (huynvk) [inline src/static/js/earn.js]: Does `allowedTokens / 1e18` mean `allowedTokens` devides by 1e18? If this division equals to 0, does it mean allowedTokens is zero? Can we replace it by `allowedTokens == 0`?

File: src/static/js/earn.js [hunk 3]
```diff
@@ -73,6 +74,88 @@ async function main() {
...
+      snobMessage(`Approve spending`, `Please approve spending first. Please check your Metamask Wallet`, `information-circle-outline`, `primary`, false, `ok`);
```
Comment (huynvk) [inline src/static/js/earn.js]: Normally, we only use "\`" when we have some embedded variable inside (e.g. \`Hello, ${username}\`). In other cases, we only use single quote (') or double quote (") depend on team convention

File: src/static/js/earn.js [hunk 4]
```diff
@@ -73,6 +74,88 @@ async function main() {
...
+    console.log(signer)
+    const STAKING_TOKEN = new ethers.Contract(S3F_ADDRESS, ERC20_ABI, signer)
+    console.log(STAKING_TOKEN)
+    const CHEF_CONTRACT = new ethers.Contract(STAKING_ADDR, STAKING_ABI, signer)
+    console.log(CHEF_CONTRACT)
+    const currentTokens = await STAKING_TOKEN.balanceOf(App.YOUR_ADDRESS)
+    console.log(currentTokens)
+    const allowedTokens = await STAKING_TOKEN.allowance(App.YOUR_ADDRESS, STAKING_ADDR)
+    console.log(allowedTokens)
+    let allow = Promise.resolve()
```
Comment (huynvk) [inline src/static/js/earn.js]: What is the purpose of `allow`?

File: src/static/js/earn.js [hunk 5]
```diff
@@ -73,6 +74,88 @@ async function main() {
...
+    if (allowedTokens / 1e18 == 0) {
+      snobMessage(`Approve spending`, `Please approve spending first. Please check your Metamask Wallet`, `information-circle-outline`, `primary`, false, `ok`);
+    } else if (currentTokens / 1e18 > 0) {
+      halfmoon.toggleModal('modal-loading')
+      allow
+        .then(async function () {
```
Comment (huynvk) [inline src/static/js/earn.js]: You use arrow function in some places, and then `function ()` in other places. Can we make it consistent?

Comment (huynvk) [reply src/static/js/earn.js]: Can we make it consistent between using `promise.then` or `async / await`? Do you need `async` function here?

File: src/static/js/earn.js [hunk 6]
```diff
@@ -73,6 +74,88 @@ async function main() {
...
+
+  const stakingContract_withdraw = async ({STAKING_ABI, STAKING_ADDR, App, AppSTAKING_CONTRACT, SNOB_TOKEN, S3F_TOKEN, renderPoolS3F}) => {
+    const signer = App.provider.getSigner()
+    console.log(signer)
+    const STAKING_CONTRACT = new ethers.Contract(STAKING_ADDR, STAKING_ABI, signer)
+    const currentTokens = await STAKING_CONTRACT.balanceOf(App.YOUR_ADDRESS)
+    let allow = Promise.resolve()
+    if (currentTokens / 1e18 > 0) {
+      halfmoon.toggleModal('modal-loading')
+      allow
```
Comment (huynvk) [inline src/static/js/earn.js]: Why do we need this `allow.then()`?

File: src/static/js/earn.js [hunk 7]
```diff
@@ -1130,50 +1207,88 @@ async function main() {
 
 
   }
-  function poolS3F(options) {
-    var eDayAPR = options.icequeen_apr;
-    var eYearAPR = options.icequeen_apr * 365;
+  function poolSNOB(options) {
+    //_print(``)
+    if (options.url) {
+      //_print(`<b>${options.pool_nickname}</b> <a href='${options.url}' target="_blank">${options.pool_name}</a>`)
```
Comment (huynvk) [inline src/static/js/earn.js]: Please remove unused code

File: src/static/js/earn.js [hunk 8]
```diff
@@ -1130,50 +1207,88 @@ async function main() {
...
-      var poolSize = `<span class="badge badge-pill font-size-12 px-5 px-sm-10 mx-5 font-weight-regular">${(options.total_staked / 1e18).toLocaleString()} S3F </span>`;
+      var eDayAPR = `${options.icequeen_apr.toFixed(2)}`;
+      var eWeekAPR = `${(options.icequeen_apr * 7).toFixed(2)}`;
+      var eYearAPR = `${(options.icequeen_apr * 365).toFixed(2)}`;
+
+
+      var combinedAprDisplay = '';
+      if (options.snowglobe_apr) {
+        let combinedAPR = options.icequeen_apr + options.snowglobe_apr
+        //_print(`Combined APR**: Day ${combinedAPR.toFixed(2)}% Week ${(combinedAPR * 7).toFixed(2)}% Year ${(combinedAPR * 365).toFixed(2)}%`)
```
Comment (huynvk) [inline src/static/js/earn.js]: Remove unused comment

File: src/static/js/earn.js [hunk 9]
```diff
@@ -1130,50 +1207,88 @@ async function main() {
...
+    availableStake = '';
     if ( options.display_amount > 0 ) {
-      availableStake = stakeUnstake(options.display_amount.toFixed(6), true, 'S3F');
+      availableStake = stakeUnstake(options.display_amount.toFixed(6), true, 'PGL');
     }
-    var availableUnstake = ''
-    if ( options.staked_pool / 1e18 > 0 ) {
-      availableUnstake = stakeUnstake((options.staked_pool / 1e18).toFixed(6), false, 'S3F');
+    availableUnstake = '';
+    if ( options.staked_pool.amount / 1e18 > 0 ) {
```
Comment (huynvk) [inline src/static/js/earn.js]: Can we replace this `options.staked_pool.amount / 1e18 > 0` by `options.staked_pool.amount > 0`

Comment (huynvk) [inline src/static/js/earn.js]: Remove trailing spaces

File: src/static/js/earn.js [hunk 10]
```diff
@@ -1183,26 +1298,25 @@ async function main() {
     if ( options.display_amount > 0 ) {
       has_options = true
       approveBtn = `<button data-btn="${options.approve}" class="btn btn-sm mx-10 approveBtn" ><ion-icon name="bag-check-outline" role="img" class="md hydrated" aria-label="bag check outline"></ion-icon> Approve</button>`;
-      stakeBtn = `<button data-btn="${options.stake}" class="btn btn-sm mx-10 btn-success stakeBtn"><ion-icon name="lock-open-outline"></ion-icon> Stake S3F</button>`;
+      stakeBtn = `<button data-btn="${options.stake}" class="btn btn-sm mx-10 btn-success stakeBtn"><ion-icon name="lock-open-outline"></ion-icon> Stake PGL</button>`;
     }
-    if ( options.staked_pool / 1e18 > 0 ) {
+    if ( options.staked_pool.amount / 1e18 > 0 ) {
```
Comment (huynvk) [inline src/static/js/earn.js]: Same here with the condition

File: src/static/js/earn.js [hunk 11]
```diff
@@ -1925,74 +1739,530 @@ async function main() {
...
+
+  function stakeUnstake(amount, stake, st){
+    return `<div class="col-sm-12 col-md-3 align-items-center text-center snob-tvl pb-10 pb-md-0">
+    <p class="m-0 font-size-12"><ion-icon name="pie-chart-outline"></ion-icon> You have</p>
+    <p class="m-0 font-size-16 font-weight-regular">${amount} ${(st?st:'sPGL')} </p>
+    <p class="m-0 font-size-12">(Available to ${(stake? 'Stake': 'Unstake')}) </p>
+    </div>`
+  }
+  
+  function poolS3F(options){
```
Comment (huynvk) [inline src/static/js/earn.js]: This is a huge function. Is it that long originally? Can we split it smaller? That's OK to leave it as is if it was already that long before

File: src/static/js/earn.js (hunk 1 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: This is from the original. These logs also remain in the production version.

File: src/static/js/earn.js (hunk 4 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: It is from the original. I see that the former dev like to create that "allow" promise for later promise chain syntax (when action submits)

File: src/static/js/earn.js (hunk 2 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: It is from the original. I didn't notice the calculations.

File: src/static/js/earn.js (hunk 3 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: It is from the original. They use this syntax for all functions to show messages.

File: src/static/js/earn.js (hunk 5 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: This snippet is from the original. The `async` is redundant but I didn't notice to remove it.

File: src/static/js/earn.js (hunk 6 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: This is from the original. As the review above, former dev likes to create the "allow" promise then chain it

File: src/static/js/earn.js (hunk 11 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: It is long originally

File: src/static/js/earn.js (hunk 9 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: This condition is original. They come from different functions. The "green" is from poolSNOB(), the "red" is from poolS3F"

File: src/static/js/earn.js (hunk 7 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: This poolSNOB function is original.

File: src/static/js/earn.js (hunk 10 above)
Comment (thongpt-codelink) [reply src/static/js/earn.js]: They come from different functions. The "green" is from poolSNOB(), the "red" is from poolS3F"

File: src/static/js/earn.js [hunk 12]
```diff
@@ -1,173 +1,1833 @@
...
+  const SNOWGLOBE_ABI = [
+    {
+      type: 'constructor',
+      stateMutability: 'nonpayable',
+      inputs: [
+        {type: 'address', name: '_token', internalType: 'address'},
+        {type: 'address', name: '_governance', internalType: 'address'},
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:30]: These changes is due to the formatter.

File: src/static/js/earn.js [hunk 13]
```diff
@@ -1,173 +1,1833 @@
...
-
-  const approveSPGLSUSHI = async function () {
+  const SUSHI_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0x14ec55f8B4642111A5aF4f5ddc56B7bE867eB6cC'
+  const SNOB_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0xB12531a2d758c7a8BF09f44FC88E646E1BF9D375'
+  const PNG_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0x6A803904b9eA0Fc982fBB077c7243c244Ae05a2d'
+  const ETH_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0x953853590b805A0E885A75A3C786D2aFfcEEA3Cf'
+  const LINK_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0x974Ef0bDA58C81F3094e124f530eF34fe70dc103'
+  const USDT_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0x74dB28797957a52a28963F424dAF2B10226ba04C'
+
+  const stakingContract_stake = async ({
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:1611]: I cleaned the unused logs and use async/await to make it consistent

File: src/static/js/earn.js [hunk 14]
```diff
@@ -339,117 +2037,118 @@ async function main() {
...
     res = await $.ajax({
       url: 'https://x-api.snowball.network/dex/0xc38f41a296a4493ff429f1238e030924a1542e50/tvl.json',
       type: 'GET',
     })
     if (res && res.pairs && res.locked) {
-
-      res.pairs.forEach( p => {
-        if ( pairmatch(p, 'usdt', 'wavax') ) {
-          pool5tvl = p.locked;
+      res.pairs.forEach(p => {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:2090]: These changes is due to the formatter

File: src/static/js/earn.js [hunk 15]
```diff
@@ -465,504 +2164,287 @@ async function main() {
...
-                ${unstakeBtn}
-                ${claimBtn}
-                </div>
+  let link_annual_apy = 100 * (1 + link_r / compounds_per_year) ** compounds_per_year - 100
+  let usdt_r = usdt_apr.yearlyAPR / 100
+  let usdt_annual_apy = 100 * (1 + usdt_r / compounds_per_year) ** compounds_per_year - 100
 
-                <div onclick="toggleDetails('${options.pool_nickname}');" class="col-sm-12 col-md-1 align-items-center text-center text-md-right snob-tvl pb-10 pb-md-0 mx-auto">
-                    <ion-icon class="pointer" alt="More Details" name="chevron-down-outline"></ion-icon>
+  const aprDisplay = (cDayAPR, cWeekAPR, cYearAPR) => {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:2171]: The aprDisplay function start from line **481** on the "red" side.

File: src/static/js/earn.js [hunk 16]
```diff
@@ -465,504 +2164,287 @@ async function main() {
...
                 </div>
             </div>
         </div>
-    </div>`;
-      $('#snob-pools-used').append(poolPrint);
-    }
+    </div>`
+  }
+
+  const calculateShare = (snowglobeContract, PAIR_ADDR, userSPGL, decimals, pool_percent) => {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:2193]: The calculateShare function start from line **503** on the "red" column

File: src/static/js/earn.js [hunk 17]
```diff
@@ -465,504 +2164,287 @@ async function main() {
...
+    stakeDisplay_6 = ret_6[2]
+    totalPoolPGL_6 = ret_6[3]
   }
-  function poolS3D(options) {
-    if (options.icequeen_apr) {
 
-      var eDayAPR = `${options.icequeen_apr.toFixed(2)}`;
-      var eWeekAPR = `${(options.icequeen_apr * 7).toFixed(2)}`;
-      var eYearAPR = `${(options.icequeen_apr * 365).toFixed(2)}`;
+  function pool(options) {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:2333]: This is 2 different function, **pool** and ** poolS3D**. The pool function start from line **609** of the "red" column

File: src/static/js/earn.js [hunk 18]
```diff
@@ -1016,197 +2500,241 @@ async function main() {
...
+            </div>
+        </div>
+    </div>`
+      $('#snob-pools-used').append(poolPrint)
+    }
   }
-  function poolS3F(options) {
-    var eDayAPR = options.icequeen_apr;
-    var eYearAPR = options.icequeen_apr * 365;
+  function poolS3D(options) {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:2627]: This is 2 different functions too. poolS3D starts from line **871** of the red column

File: src/static/js/earn.js [hunk 19]
```diff
@@ -1366,115 +2910,110 @@ async function main() {
...
       </div>
-  </div>`;
-      $('#snob-pools-new').append(poolPrint);
+  </div>`
+      $('#snob-pools-new').append(poolPrint)
     }
-
-
   }
   function poolSNOB(options) {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:2917]: I cleaned the unused codes

File: src/static/js/earn.js [hunk 20]
```diff
@@ -1639,46 +3196,31 @@ async function main() {
...
-    unstake: 'withdrawPool8',
-    claim: 'claimPool8',
-    icequeen_apr: pool8APR,
-    snowglobe_apr: null,
-    tvl_display: pool8tvlDisplay,
-    total_pgl: null,
-    pool_share_display: poolShareDisplay_8,
-    pool_share_display_pgl: '',
-    stake_display: ''
+  await renderPoolS3F({
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:3211]: I replaced the **poolS3F** using this function. This function is declared later using the old function declaration syntax to hoist.

File: src/static/js/earn.js [hunk 21]
```diff
@@ -1860,140 +3414,625 @@ async function main() {
...
-        alert('Oops something went wrong. Try refreshing the page.');
+  function stakeUnstake(amount, stake, st) {
+    return `<div class="col-sm-12 col-md-3 align-items-center text-center snob-tvl pb-10 pb-md-0">
+    <p class="m-0 font-size-12"><ion-icon name="pie-chart-outline"></ion-icon> You have</p>
+    <p class="m-0 font-size-16 font-weight-regular">${amount} ${st ? st : 'sPGL'} </p>
+    <p class="m-0 font-size-12">(Available to ${stake ? 'Stake' : 'Unstake'}) </p>
+    </div>`
+  }
+
+  function poolS3F(options) {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:3560]: The **poolS3F** is an original function. It is long, but I don't know the business well enough to change.

File: src/static/js/earn.js [hunk 22]
```diff
@@ -1860,140 +3414,625 @@ async function main() {
...
+  
+                    ${availableStake}
+                    
+                    ${availableUnstake}
+                </div>
+            </div>
+        </div>
+    </div>`
 
-  $(".claimBtn").click(function(){
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:1898]: This should compare with line **3454** on the "green" section

File: src/static/js/earn.js [hunk 23]
```diff
@@ -2299,90 +4519,37 @@ const stakingContract_approve = async function (chefAbi, chefAddress, stakeToken
...
         halfmoon.toggleModal('modal-loading')
         return App.provider.waitForTransaction(t.hash)
       })
-      .catch(function () {
-        halfmoon.toggleModal('modal-loading')
-        snobMessage(`Connecting to metamask`, `Approval failed . Please check your Metamask Wallet`, `close-circle-outline`, `danger`, false, `ok`, 4000);
-      })
-  }
-}
-const stakingContract_stake = async function (chefAbi, chefAddress, stakeTokenAddr, App) {
```
Comment (thongpt-codelink) [inline src/static/js/earn.js:2315]: The **stakingContract_stake** is moved to line **1611** in the "green" column

**Pull request closed without merging**